### PR TITLE
Livecheck: Replace OpenURI#open with Curl

### PR DESCRIPTION
**Pull request by samford**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR migrates `brew livecheck` from `open-uri` to `Utils::Curl`, as `Curl` is used throughout Homebrew/brew for network requests and livecheck was the only place where `open-uri` was used (outside of a vendored gem).

## Notable Changes

* Replaces `open-uri` in `PageMatch` with `#curl_with_workarounds`. Most other strategies use `PageMatch#find_versions` internally, so this is one of the two strategies that used `open-uri` (the other being `Xorg`). For what it's worth, some of the code in `#page_matches`/`#page_content` was adapted from methods in `Utils::Curl` (e.g., `#curl_http_content_headers_and_checksum`).
* Adds a `user_agent` parameter to `Strategy#page_content`, so we can pass a symbol to `curl_args` to configure `curl`'s user agent. The `user_agent` parameter could be used in strategies, if we ever encounter a page that requires a different user agent string to work. Similarly, we could modify the `livecheck` DSL to allow user agent configuration but that's something we should defer until it's actually necessary for a given formula/cask's check to function. That said, I imagine these cases will be quite rare but adding this parameter (and passing it in the `curl` args) takes almost no work, so I don't see any adding it.

## Closing

This PR is currently blocked while I troubleshoot a hang that occurs on this `curl` version of livecheck. Basically, either `curl` or `SystemCommand` is hanging in a random place on longer livecheck runs (e.g., `brew livecheck --tap homebrew/core`) and I haven't yet determined the cause of the hang. This is a crucial bug that needs to be addressed before this can be merged.

### CODE REVIEW

**Comment by samford** [comment]
The latest push moves `#page_content` out of `PageMatch` and into `Strategy`, to better align with #9529.

**Comment by reitermarkus** [comment]
Is there a specific reason we are not using `curl` for `page_content`?

**Comment by MikeMcQuaid** [comment]
> Is there a specific reason we are not using `curl` for `page_content`, i.e. `curl_output`?

I'd like to strongly suggest we use `curl` everywhere for consistently.

**Comment by reitermarkus** [comment]
> I'd like to strongly suggest we use `curl` everywhere for consistently.

I also already encountered some URLs which simply won't work using `URI.open` but work fine with `curl`.

**Comment by samford** [comment]
A couple months back, I reimplemented `#page_matches` (`#page_content` here) using `#curl_with_workarounds` as a solution to this particular redirection issue but I didn't get around to creating a PR. The work is complete, so I'll take a pass at it in a moment to bring it up to date with what I'm doing in this PR.

The short version of why we're not using `curl` is that there were some technical hurdles to make it work properly within the context of livecheck. Basically, `curl`'s output can make a mess of our debug and verbose JSON output and it's necessary to make some changes that allow us to avoid this (in `SystemCommand` and `Utils::Curl`). These changes are part of the aforementioned reimplementation, so that's already resolved.

I was initially concerned that using `curl` would be slower but, from some very basic testing, I haven't see any notable difference that couldn't be attributed to normal variations between two runs. Using `curl` does appear to double the CPU usage of livecheck (compared to using `open-uri`) but I imagine that being consistent about using `curl` throughout `brew` outweighs this.

There's been support for reworking livecheck to use `curl` (instead of `open-uri`) outside of this PR. If the prevailing view is "livecheck should use `curl`, no question", I'll update this PR to use my `curl` implementation. I originally worked on the `open-uri` and `curl` implementations at the same time (for the sake of comparison), so the `curl` version is a drop-in replacement.

**Comment by MikeMcQuaid** [comment]
> If the prevailing view is "livecheck should use `curl`, no question", I'll update this PR to use my `curl` implementation.

Yes, ideally. It also means that e.g. proxies only need configured in a single way.

**Comment by samford** [comment]
Gotcha, I'll push the `curl` version sometime today after I've brought it up to date and tested it.

Standardizing on using `curl` will allow us to address an issue in #9529 as well, so it makes sense to me.

**Review by samford — COMMENTED** [review]
I think this is ready for review now. I reworked the aforementioned curl version of this, pushed it, and revised the opening PR text explaining everything.

I've commented on a couple areas here where I had questions or could use another set of eyes.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [inline Library/Homebrew/livecheck/strategy.rb:117]:
```suggestion
          location = headers[/^Location:\s*(.*)$/i, 1]&.chomp
```

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [inline Library/Homebrew/livecheck/strategy.rb:120]:
```suggestion
```

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb:124]: I think this same logic, which also handles multiple redirects, also exists in `download_strategies.rb`.

File: Library/Homebrew/system_command.rb
Comment (reitermarkus) [reply Library/Homebrew/system_command.rb:85]: Actually, I think `debug` and `verbose` should be `nil`, given we check for it in `debug?` and `verbose?`.

**Review by reitermarkus — CHANGES_REQUESTED** [review]

File: Library/Homebrew/livecheck/strategy.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy.rb:124]: Perhaps extracting both to `utils/curl` would be good?

File: Library/Homebrew/system_command.rb
Comment (samford) [reply Library/Homebrew/system_command.rb]: I was wondering about that but I forgot to mention it. I added a commit to set them to `nil` by default. I'm not very experienced with Sorbet yet but should the type signatures be something like `T.nilable(T::Boolean)`?

File: Library/Homebrew/livecheck/strategy.rb
Comment (samford) [reply Library/Homebrew/livecheck/strategy.rb]: There are similarities in what's happening in these three locations, so I tried my hand at creating a method that can handle these use cases (`#response_status_code_and_location`). I welcome feedback on the shape it takes, where it should exist, etc. but the current code is functional.

---

Basically, the work that I saw repeated in these areas was:

* Separating response headers from the body.
* Iterating through the headers for each of the responses to extract information.

The main functional difference between the approaches is whether the final `location` URL should be made absolute:

* [`Utils::Curl#curl_http_content_headers_and_checksum`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/curl.rb#L244-L250) seems to be geared toward returning headers in their original state, so it doesn't modify the `location` header URL.
* [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441) modifies the `location` header URL to make it absolute (using the original URL). This is more advanced than the simple `URI#join` statement I had in `Strategy#page_content`.

With this in mind, I made a method for separating the headers and body in cURL response text (`#response_headers_and_body`). This should work properly when the response contains both headers and body, only headers, and only the body.

If you need to work with the separated `headers`/`body` yourself, then you can execute this method before calling `#response_status_code_and_location` to extract the status_code and/or final location URL. If you don't need to separate the `headers`/`body`, then you can simply pass `#response_status_code_and_location` the response text and it will call `#response_headers_and_body` internally to format the separate the headers into an array.

---

Once this had gone through review, I'll add tests for these methods.

File: Library/Homebrew/utils/curl.rb
Comment (reitermarkus) [inline Library/Homebrew/utils/curl.rb:269]: I think `dirname` is actually wrong when the path ends with a slash, e.g.

```
http://example.org/subdir/ --> ./file1 --> http://example.org/file1
``` 

instead of 

```
http://example.org/subdir/ --> ./file1 --> http://example.org/subdir/file1
``` 

since `Pathname("subdir/").dirname` and `Pathname("subdir").dirname` are the same.

Comment (samford) [reply Library/Homebrew/utils/curl.rb:269]: Nice catch. I simply copied over the existing logic from  [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441), so any preexisting issues remain.

I'll start working on some tests to exercise this, as I hadn't yet tested the existing code that makes URLs absolute.

Comment (reitermarkus) [reply Library/Homebrew/utils/curl.rb:269]: I think in the existing code we only use it as a fallback for getting the file name so this bug would never manifest itself.

Comment (samford) [reply Library/Homebrew/utils/curl.rb]: With tests in place for these relative URL cases, I initially created a fix for the `./` case but wasn't quite satisfied with it. I tested `URI#join` and it correctly handles all of these cases, so I used it to replace the absolute URL code that I copied from [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441).

The end result is a simple one-liner that uses `URI#join`. I imagine `URI#join` is more correct/reliable than the code from `CurlDownloadStrategy`, so it seems like a better choice overall.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [inline Library/Homebrew/livecheck/strategy.rb:106]: Do you mean 

```suggestion
        unless result.status.success?
```

?

Comment (samford) [reply Library/Homebrew/livecheck/strategy.rb]: Yes, that's certainly a better representation of what I was aiming to achieve and I've confirmed that we still identify errors as expected. I'm not sure what I was thinking when I originally wrote that but I've been staying up too late working on this recently, ahah.

This change is included in the latest push.

**Comment by samford** [comment]
This latest push resolves a bug in `Utils::Curl#response_headers_and_body` that I discovered in testing. Basically, using `response_text.rpartition("\r\n\r\n")` to separate the headers/body was overly optimistic, as body content containing `\r\n\r\n` led to some of the body being treated as a header.

I addressed this by being more strict about identifying and separating the response headers from the body. I've expanded the tests to include related cases, ensuring they correctly failed before and that the new code passes.

---

I've also added `--connect-timeout 15` and `--max-time 30` to the `curl_args` in `Strategy#page_content`, as `livecheck` has continued to hang at times due to `curl`.

This wasn't enough to resolve the hanging situation and the default maximum retry time may also be a factor, since `#curl_args` always includes `--retry`. We don't need retries in livecheck, so I modified `#curl_args` to include an option to omit `--retry` (though it continues to be included by default).

Unfortunately, even omitting `--retry` in livecheck doesn't seem to avoid the situation where `curl` will sometimes endlessly hang somewhere in the middle of a long run. Just for good measure, I tried added `--retry-max-time 30` (in addition to all the above) and it can still run into an endless hang.

I've been doing homebrew/core runs on an otherwise idle machine and the issue still regularly arises.. Some runs are fine and others hit a brick wall somewhere along the way (but seemingly never in the same place). When it hangs, there's no choice but to abandon the run and start over. Considering that a full homebrew/core run takes awhile and `--json` doesn't print data until the end, hanging in the middle of a run and losing all that progress is a pain.

I'm going to have to dig into this deeper to better understand what's going wrong. As it stands, this issue is the main thing holding up this PR, as I don't feel we should merge this until it's resolved. Let me know if anyone has any idea of what could be causing the `curl` hang or ways in which we might resolve it.

**Comment by reitermarkus** [comment]
Is it this bug https://github.com/Homebrew/brew/issues/6648 which was fixed by https://github.com/Homebrew/brew/pull/6655 but only really fixed by https://github.com/Homebrew/brew/pull/9496?

File: Library/Homebrew/livecheck/strategy/xorg.rb
Comment (reitermarkus) [inline Library/Homebrew/livecheck/strategy/xorg.rb:86]: Where is this cached content coming from?

Comment (samford) [reply Library/Homebrew/livecheck/strategy/xorg.rb:86]: `Xorg` formulae/casks basically end up checking a small number of directory listing pages that contain a bunch of different software. Rather than checking the same pages over and over, the strategy saves the HTML in `@page_data`, using the page URL as the key. This way, we only fetch the page once. It also makes subsequent checks [using the same page] very fast, so running through homebrew-xorg is faster as a result.

This naive approach is fine within this limited context but I have something more sophisticated in the works for future livecheck-wide caching. This `Xorg` logic is just something I threw together while refactoring the strategy as part of migrating it from homebrew-xorg (when some X.org formulae were being migrated to homebrew/core).

**Comment by samford** [comment]
> Is it this bug #6648 which was almost fixed by #6655 but only really fixed by #9496?

Thanks for the references.

From my testing today, the only thing I can say with any certainty is that when `curl` hangs, the command is run but we don't get to the point where response data would normally appear. If you enable all the output options (`print_stdout`, `print_stderr`, `debug`, `verbose`), you would normally see the `curl` command printed, followed by the HTTP response headers and body content. When this hangs, we only see the `curl` command (no HTTP response headers/body).

I'll test this more tomorrow, in hopes of understanding if this is related to `SystemCommand`, `curl` itself, or something else entirely.

**Comment by MikeMcQuaid** [comment]
@samford What's the latest here?

**Comment by samford** [comment]
> What's the latest here?

I tried to rebase and push this when it was marked as stale but the link between the PR and branch seems to be broken, as it doesn't pick up any changes to my branch. The branch link at the top isn't a link anymore (though [the branch still exists in my fork](https://github.com/samford/brew/tree/pagematch-handle-redirections)), so it's like the PR and the branch have been decoupled. I may have to close this PR and create a new one if there isn't an easier way to fix this.

I periodically revisit this to rebase and test it and the previously-described hang was still happening a few weeks ago. I'll try to revisit my previous tests over the next few days, as it's been a while since I've tinkered with this in earnest and I forget what did and didn't work. I'll report my findings once I've finished retesting this, as someone else may have some insight into the issue that I'm missing.

When I last left this, my hypothesis was that the issue was in `SystemCommand` but none of the changes I made resolved the issue and I had hit a dead end. The basic idea is that when we encounter a hang, the `curl` process has finished/exited but `Open3` is still running. `SystemCommand` uses `popen3` and it can hang if you don't close a stream, so this may play a part in this issue.

**Comment by MikeMcQuaid** [comment]
> When I last left this, my hypothesis was that the issue was in `SystemCommand` but none of the changes I made resolved the issue and I had hit a dead end. The basic idea is that when we encounter a hang, the `curl` process has finished/exited but `Open3` is still running. `SystemCommand` uses `popen3` and it can hang if you don't close a stream, so this may play a part in this issue.

@samford I think it may be worth landing this change regardless and then trying to get more eyes on it post-merge. As-is there are bugs that will be fixed by this change and it's really something I should have spotted before the livecheck merge and would have required before that (as, as-is, all proxy settings in Homebrew won't work properly and certificates work inconsistently compared to all the other `curl` usage).

**Comment by samford** [comment]
I've been testing this over the past couple days and, surprisingly, I haven't encountered the hang yet. It typically appears pretty frequently but there are also periods where it doesn't appear at all, so it could just be a fluke. I would like to continue testing this for another day to see what it does.

There are some areas that need to be tweaked before we consider merging this but, unfortunately, pushing to this branch doesn't update the PR. It seems like I'm going to have to close this PR and create a new one, unless you know of another way to resolve this issue.

**Comment by MikeMcQuaid** [comment]
> It seems like I'm going to have to close this PR and create a new one, unless you know of another way to resolve this issue.

Yes, I think so.

> I've been testing this over the past couple days and, surprisingly, I haven't encountered the hang yet. It typically appears pretty frequently but there are also periods where it doesn't appear at all, so it could just be a fluke. I would like to continue testing this for another day to see what it does.

That's good.

**Comment by samford** [comment]
Closing this in favor of opening a new PR that's not broken.

**Pull request closed without merging**
